### PR TITLE
Add Makefile.js and eslint

### PR DESCRIPTION
**Pull request by soda0289**
We currently have the lint target disabled until the babel is removed. The test run but depend on babel-core/register. When https://github.com/eslint/eslint-scope/pull/14 is merged I will turn on the linter.

### CODE REVIEW

**Comment by JamesHenry** [comment]
Great stuff, thanks! I'll let you know as soon as #14 is merged

**Comment by JamesHenry** [comment]
@soda0289 The build failed: `istanbul: not found`

**Comment by soda0289** [comment]
Do we want code coverage? I will just remove it for now.

**Comment by JamesHenry** [comment]
Ideally, yes, as that would match up with our other projects. Is it not just a matter of adding istanbul as a dependency?

**Comment by soda0289** [comment]
Hmm I'm not sure how travis works. I can add it as dependency and call by the node_modues path. Let me take a look at how eslint does it

**Comment by soda0289** [comment]
I got code coverage working just going to fix all these linting errors. Give me an hour.

**Comment by soda0289** [comment]
It's going to take a little longer to finish fixing all the linting rules. I have pushed a WIP commit but will finish it later this evening.

**Comment by soda0289** [comment]
@JamesHenry Should be good to go. I disabled some linting rules in the tests, but mostly had to add JSDoc comments and make minor tweaks to get everything passing.

File: src/scope.js [hunk 1]
```diff
@@ -400,19 +422,19 @@ class Scope {
     }
 
     __isClosed() {
-        return this.__left === null;
+        return this.__left === null || this.__left === undefined;
```
Comment (JamesHenry) [inline src/scope.js]: This is a code change, I have no idea if it is significant, but thought I should point it out as it is not necessary for the PR

Comment (soda0289) [inline src/scope.js]: They had some parts of the code that had != null and == null and I changed those to triple equals with undefined. I can change this one back.

Comment (JamesHenry) [inline src/scope.js]: Yes, I saw, and so you did not change that meaning of the code in those cases. Here you did, that's why I am pointing it out

Comment (JamesHenry) [inline src/scope.js]: As I say, I have no idea what significance that has, but it's probably worth pointing out

Comment (soda0289) [inline src/scope.js]: I changed it back. Although the tests pass I dont want to risk introducing new bugs.

File: src/pattern-visitor.js [hunk 2]
```diff
@@ -56,7 +61,7 @@ class PatternVisitor extends esrecurse.Visitor {
         const lastRestElement = getLast(this.restElements);
         this.callback(pattern, {
             topLevel: pattern === this.rootPattern,
-            rest: lastRestElement != null && lastRestElement.argument === pattern,
+            rest: lastRestElement !== null && lastRestElement.argument === pattern,
```
Comment (JamesHenry) [inline src/pattern-visitor.js]: Ah here's another code change

Comment (soda0289) [inline src/pattern-visitor.js]: good catch will fix. Thanks

File: src/referencer.js [hunk 3]
```diff
@@ -23,24 +23,34 @@
...
+ * @returns {void}
+ */
 function traverseIdentifierInPattern(options, rootPattern, referencer, callback) {
     // Call the callback at left hand identifier nodes, and Collect right hand nodes.
     var visitor = new PatternVisitor(options, rootPattern, callback);
     visitor.visit(rootPattern);
 
     // Process the right hand nodes recursively.
-    if (referencer != null) {
+    if (referencer !== null) {
```
Comment (JamesHenry) [inline src/referencer.js]: Another one

Comment (soda0289) [inline src/referencer.js]: Got it. Thanks

**Comment by soda0289** [comment]
I pushed the changes. Think that's all of them. Thanks for the review

**Comment by mikesherov** [comment]
LGTM, only thing I'd say is to add an .eslintrc file to the test directory that extends the main .eslintrc file and allows describe and it as globals.

**Comment by soda0289** [comment]
@mikesherov Great Idea. I can set environment to mocha in the test directory and that should handle all the globals. Thanks

**Comment by JamesHenry** [comment]
Is this finalised now @soda0289?

**Comment by soda0289** [comment]
Yup should be good.

**Pull request merged**